schema/avro/union.go

```diff
@@ -202,59 +207,80 @@ func (r unionResolver) afterUnmarshalArraySubstitutions(val any, substitutions [
 }
 
 func (r unionResolver) afterUnmarshalNullUnionSubstitutions(val any, substitutions []substitution) ([]substitution, error) {
-	for _, p := range r.nullUnionPaths {
-		// first collect all values that are nullable
-		var maps []map[string]any
-		err := traverseValue(val, p, true, func(v any) {
-			switch v := v.(type) {
-			case map[string]any:
-				maps = append(maps, v)
-			case *map[string]any:
-				maps = append(maps, *v)
-			case *opencdc.StructuredData:
-				maps = append(maps, *v)
-			}
-		})
+	for _, nullUnionPath := range r.nullUnionPaths {
+		// first collect all parents that contain a value that is nullable
+		parentMaps, err := r.collectParentsForNullUnionPath(val, nullUnionPath)
 		if err != nil {
 			return nil, err
 		}
 
-		// Loop through collected maps and collect all substitutions. These maps
-		// contain values encoded as maps with a single key:value pair, where
-		// key is the type name (e.g. {"int":1}). We want to replace all these
-		// maps with the actual value (e.g. 1).
-		// We don't replace them in the loop, because we want to make sure all
-		// maps actually contain only 1 value.
```

Comment:
Nitpick: I find this comment very useful to understand what we're doing, I propose leaving it in the code.

Reply:
Yup, it definitely. I removed it because it's present in another place(s), but it doesn't hurt having a reminder here too.:)